File: backend/ai/gemini_client.py
"""
Gemini AI Client - Integration with Google's Gemini AI for intelligent analysis
"""

import google.generativeai as genai
import logging
from typing import Dict, List, Optional
import json

from config.settings import settings

logger = logging.getLogger(__name__)


class GeminiClient:
    """Client for Gemini AI integration"""
    
    def __init__(self):
        self.model = None
        self.ready = False
        self.request_count = 0
        
    def initialize(self):
        """Initialize Gemini AI client"""
        if not settings.GEMINI_API_KEY:
            logger.warning("Gemini API key not configured")
            return False
        
        try:
            genai.configure(api_key=settings.GEMINI_API_KEY)
            self.model = genai.GenerativeModel(settings.GEMINI_MODEL)
            self.ready = True
            logger.info(f"Gemini AI initialized with model: {settings.GEMINI_MODEL}")
            return True
        except Exception as e:
            logger.error(f"Failed to initialize Gemini: {e}")
            return False
    
    async def analyze_pattern(
        self,
        features: Dict,
        patterns: List[str],
        candles: List[Dict],
        recent_performance: Optional[Dict] = None
    ) -> Optional[Dict]:
        """
        Analyze trading pattern using Gemini AI
        
        Args:
            features: Extracted features
            patterns: Detected patterns
            candles: Recent candle data
            recent_performance: Recent win/loss statistics
            
        Returns:
            AI analysis with insights and confidence adjustment
        """
        if not self.ready:
            return None
        
        try:
            # Prepare context for AI
            context = self._prepare_context(features, patterns, candles, recent_performance)
            
            # Create prompt
            prompt = self._create_analysis_prompt(context)
            
            # Get AI response
            response = self.model.generate_content(prompt)
            
            # Parse response
            analysis = self._parse_response(response.text)
            
            self.request_count += 1
            
            return analysis
            
        except Exception as e:
            logger.error(f"Gemini analysis failed: {e}")
            return None
    
    def _prepare_context(
        self,
        features: Dict,
        patterns: List[str],
        candles: List[Dict],
        recent_performance: Optional[Dict]
    ) -> Dict:
        """Prepare context data for AI"""
        # Get last 5 candles
        recent_candles = candles[-5:] if len(candles) >= 5 else candles
        
        # Simplify candle data
        candle_summary = []
        for c in recent_candles:
            direction = "🟢 UP" if c['close'] > c['open'] else "🔴 DOWN"
            body_size = abs(c['close'] - c['open'])
            candle_summary.append(f"{direction} (body: {body_size:.5f})")
        
        context = {
            'patterns_detected': patterns if patterns else ['none'],
            'recent_candles': candle_summary,
            'trend': self._determine_trend(features),
            'volatility': features.get('volatility_ratio', 1.0),
            'near_support': features.get('near_support', 0) == 1,
            'near_resistance': features.get('near_resistance', 0) == 1,
        }
        
        if recent_performance:
            context['recent_performance'] = recent_performance
        
        return context
    
    def _determine_trend(self, features: Dict) -> str:
        """Determine overall trend from features"""
        short_slope = features.get('short_term_slope', 0)
        medium_slope = features.get('medium_term_slope', 0)
        long_slope = features.get('long_term_slope', 0)
        
        avg_slope = (short_slope + medium_slope + long_slope) / 3
        
        if avg_slope > 0.0001:
            return "UPTREND"
        elif avg_slope < -0.0001:
            return "DOWNTREND"
        else:
            return "SIDEWAYS"
    
    def _create_analysis_prompt(self, context: Dict) -> str:
        """Create prompt for Gemini AI"""
        prompt = f"""You are an expert binary options trading analyst. Analyze the following market data and provide insights.

**Current Market Context:**
- Detected Patterns: {', '.join(context['patterns_detected'])}
- Recent Candles: {' → '.join(context['recent_candles'])}
- Overall Trend: {context['trend']}
- Volatility: {'HIGH' if context['volatility'] > 1.5 else 'NORMAL' if context['volatility'] > 0.8 else 'LOW'}
- Near Support: {'YES' if context['near_support'] else 'NO'}
- Near Resistance: {'YES' if context['near_resistance'] else 'NO'}
"""

        if 'recent_performance' in context:
            perf = context['recent_performance']
            prompt += f"\n**Recent Performance:**\n- Win Rate: {perf.get('win_rate', 0):.1%}\n- Last 10 Trades: {perf.get('last_10', 'N/A')}\n"

        prompt += """
**Your Task:**
1. Analyze if this looks like broker manipulation (sudden reversals, suspicious patterns)
2. Predict the next likely move (UP or DOWN)
3. Provide confidence level (0.0 to 1.0)
4. Give brief reasoning

**Respond in JSON format:**
{
    "prediction": "UP" or "DOWN",
    "confidence": 0.0-1.0,
    "reasoning": "brief explanation",
    "manipulation_detected": true/false,
    "manipulation_reason": "explanation if detected",
    "risk_level": "LOW", "MEDIUM", or "HIGH"
}

Be concise and data-driven. Focus on technical analysis."""

        return prompt
    
    def _parse_response(self, response_text: str) -> Dict:
        """Parse AI response"""
        try:
            # Try to extract JSON from response
            start = response_text.find('{')
            end = response_text.rfind('}') + 1
            
            if start >= 0 and end > start:
                json_str = response_text[start:end]
                analysis = json.loads(json_str)
                
                # Validate required fields
                if 'prediction' in analysis and 'confidence' in analysis:
                    return analysis
            
            logger.warning("Could not parse AI response as JSON")
            return None
            
        except Exception as e:
            logger.error(f"Failed to parse AI response: {e}")
            return None
    
    async def detect_broker_manipulation(
        self,
        predictions_history: List[Dict]
    ) -> Dict:
        """
        Analyze prediction history to detect broker manipulation
        
        Args:
            predictions_history: List of past predictions with outcomes
            
        Returns:
            Manipulation analysis
        """
        if not self.ready or len(predictions_history) < 20:
            return {'manipulation_detected': False, 'confidence': 0.0}
        
        try:
            # Prepare statistics
            total = len(predictions_history)
            correct = sum(1 for p in predictions_history if p.get('was_correct'))
            win_rate = correct / total if total > 0 else 0
            
            # Check for suspicious patterns
            recent_losses = 0
            for p in predictions_history[-10:]:
                if not p.get('was_correct'):
                    recent_losses += 1
            
            prompt = f"""Analyze this binary trading performance for broker manipulation:

**Statistics:**
- Total Predictions: {total}
- Win Rate: {win_rate:.1%}
- Recent Losses (last 10): {recent_losses}/10

**Prediction History (last 20):**
"""
            for i, p in enumerate(predictions_history[-20:]):
                result = "✓" if p.get('was_correct') else "✗"
                prompt += f"\n{i+1}. {p.get('prediction')} ({p.get('confidence'):.0%}) - {result}"
            
            prompt += """

**Detect:**
1. Is the win rate suspiciously low despite high confidence?
2. Are losses clustering in unusual patterns?
3. Do outcomes seem manipulated?

**Respond in JSON:**
{
    "manipulation_detected": true/false,
    "confidence": 0.0-1.0,
    "reasoning": "explanation",
    "recommendation": "advice for trader"
}"""

            response = self.model.generate_content(prompt)
            return self._parse_response(response.text)
            
        except Exception as e:
            logger.error(f"Manipulation detection failed: {e}")
            return {'manipulation_detected': False, 'confidence': 0.0}
    
    def is_ready(self) -> bool:
        """Check if client is ready"""
        return self.ready
    
    def get_stats(self) -> Dict:
        """Get usage statistics"""
        return {
            'ready': self.ready,
            'model': settings.GEMINI_MODEL,
            'request_count': self.request_count
        }

File: extension/ai/gemini-client.js
/**
 * Gemini AI Client - Integration with Google's Gemini API
 * Provides intelligent analysis and insights for trading predictions
 */

import { logger } from '../utils/logger.js';
import { StorageUtils } from '../utils/storage.js';
import { CONFIG } from '../config.js';

export class GeminiClient {
    constructor() {
        this.apiKey = null;
        this.model = CONFIG.GEMINI.defaultModel;
        this.requestCount = 0;
        this.lastRequestTime = 0;
    }

    /**
     * Initialize client with API key
     */
    async init() {
        this.apiKey = await StorageUtils.getGeminiApiKey();
        const settings = await StorageUtils.getSettings();
        this.model = settings.selectedGeminiModel || CONFIG.GEMINI.defaultModel;

        if (!this.apiKey) {
            logger.warn('Gemini API key not configured');
            return false;
        }

        logger.info('Gemini client initialized', { model: this.model });
        return true;
    }

    /**
     * Check if client is ready
     */
    isReady() {
        return !!this.apiKey;
    }

    /**
     * Make API request with rate limiting
     */
    async makeRequest(prompt, options = {}) {
        if (!this.isReady()) {
            logger.warn('Gemini client not ready');
            return null;
        }

        // Rate limiting
        await this.enforceRateLimit();

        const url = `https://generativelanguage.googleapis.com/v1beta/models/${this.model}:generateContent?key=${this.apiKey}`;

        const requestBody = {
            contents: [{
                parts: [{
                    text: prompt
                }]
            }],
            generationConfig: {
                temperature: options.temperature || 0.7,
                topK: options.topK || 40,
                topP: options.topP || 0.95,
                maxOutputTokens: options.maxOutputTokens || 1024
            }
        };

        try {
            const response = await this.fetchWithRetry(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(requestBody)
            });

            if (!response.ok) {
                throw new Error(`API request failed: ${response.status}`);
            }

            const data = await response.json();

            if (data.candidates && data.candidates.length > 0) {
                const text = data.candidates[0].content.parts[0].text;
                logger.debug('Gemini response received', { length: text.length });
                return text;
            }

            logger.warn('No response from Gemini');
            return null;
        } catch (error) {
            logger.error('Gemini API request failed', error);
            return null;
        }
    }

    /**
     * Fetch with retry logic
     */
    async fetchWithRetry(url, options, retries = CONFIG.GEMINI.maxRetries) {
        for (let i = 0; i < retries; i++) {
            try {
                const controller = new AbortController();
                const timeout = setTimeout(() => controller.abort(), CONFIG.GEMINI.timeout);

                const response = await fetch(url, {
                    ...options,
                    signal: controller.signal
                });

                clearTimeout(timeout);
                return response;
            } catch (error) {
                if (i === retries - 1) throw error;
                logger.warn(`Retry ${i + 1}/${retries}`, error);
                await new Promise(resolve => setTimeout(resolve, 1000 * (i + 1)));
            }
        }
    }

    /**
     * Enforce rate limiting
     */
    async enforceRateLimit() {
        const now = Date.now();
        const timeSinceLastRequest = now - this.lastRequestTime;
        const minInterval = 60000 / CONFIG.GEMINI.rateLimit.requestsPerMinute;

        if (timeSinceLastRequest < minInterval) {
            const waitTime = minInterval - timeSinceLastRequest;
            logger.debug(`Rate limiting: waiting ${waitTime}ms`);
            await new Promise(resolve => setTimeout(resolve, waitTime));
        }

        this.lastRequestTime = Date.now();
        this.requestCount++;
    }

    /**
     * Analyze pattern with AI
     */
    async analyzePattern(features, patterns, candles) {
        const prompt = this.buildPatternAnalysisPrompt(features, patterns, candles);
        const response = await this.makeRequest(prompt, { temperature: 0.5 });

        if (!response) return null;

        try {
            // Parse JSON response
            const analysis = JSON.parse(response);
            return analysis;
        } catch (error) {
            logger.warn('Failed to parse Gemini response as JSON', error);
            return { analysis: response };
        }
    }

    /**
     * Build pattern analysis prompt
     */
    buildPatternAnalysisPrompt(features, patterns, candles) {
        const last5 = candles.slice(-5).map(c => ({
            open: c.open.toFixed(2),
            close: c.close.toFixed(2),
            high: c.high.toFixed(2),
            low: c.low.toFixed(2),
            direction: c.close > c.open ? 'bullish' : 'bearish'
        }));

        return `You are an expert binary trading analyst. Analyze this candlestick pattern and provide insights.

**Recent Candles (last 5):**
${JSON.stringify(last5, null, 2)}

**Detected Patterns:**
${patterns.join(', ') || 'None'}

**Key Features:**
- Short-term trend slope: ${features.shortTermSlope?.toFixed(4)}
- Medium-term trend slope: ${features.mediumTermSlope?.toFixed(4)}
- Volatility ratio: ${features.volatilityRatio?.toFixed(2)}
- Consecutive bullish: ${features.consecutiveBullish}
- Consecutive bearish: ${features.consecutiveBearish}
- Near support: ${features.nearSupport ? 'Yes' : 'No'}
- Near resistance: ${features.nearResistance ? 'Yes' : 'No'}

Provide a JSON response with:
{
  "direction": "UP" or "DOWN",
  "confidence": 0-1,
  "reasoning": "brief explanation",
  "riskFactors": ["factor1", "factor2"],
  "keyIndicators": ["indicator1", "indicator2"]
}`;
    }

    /**
     * Detect broker manipulation
     */
    async detectManipulation(predictions, stats) {
        const prompt = this.buildManipulationDetectionPrompt(predictions, stats);
        const response = await this.makeRequest(prompt, { temperature: 0.3 });

        if (!response) return null;

        try {
            return JSON.parse(response);
        } catch (error) {
            return { suspicious: false, analysis: response };
        }
    }

    /**
     * Build manipulation detection prompt
     */
    buildManipulationDetectionPrompt(predictions, stats) {
        return `You are a fraud detection expert analyzing binary trading patterns for broker manipulation.

**Statistics:**
- Total predictions: ${stats.totalPredictions}
- Win rate: ${(stats.winRate * 100).toFixed(1)}%
- Recent win rate (last 20): ${(stats.recentWinRate * 100).toFixed(1)}%
- Average confidence: ${(stats.avgConfidence * 100).toFixed(1)}%

**Recent Predictions (last 10):**
${predictions.slice(-10).map(p =>
            `- Predicted: ${p.prediction}, Actual: ${p.actualOutcome}, Confidence: ${(p.confidence * 100).toFixed(1)}%`
        ).join('\n')}

Analyze for suspicious patterns that might indicate broker manipulation:
- Unusual loss clustering
- Win rate significantly below expected
- High confidence predictions failing consistently
- Patterns that suggest price manipulation

Provide JSON response:
{
  "suspicious": true/false,
  "confidence": 0-1,
  "indicators": ["indicator1", "indicator2"],
  "recommendation": "action to take"
}`;
    }

    /**
     * Get trading advice
     */
    async getTradingAdvice(prediction, context) {
        const prompt = `As a binary trading advisor, provide advice for this prediction:

**Prediction:** ${prediction.prediction} (${(prediction.confidence * 100).toFixed(1)}% confidence)
**Patterns:** ${prediction.patterns?.join(', ') || 'None'}
**Method:** ${prediction.method}

**Context:**
- Recent win rate: ${(context.winRate * 100).toFixed(1)}%
- Consecutive losses: ${context.consecutiveLosses}
- Today's trades: ${context.todayTrades}

Should the user take this trade? Provide JSON:
{
  "recommendation": "TAKE" or "SKIP",
  "reasoning": "brief explanation",
  "riskLevel": "LOW", "MEDIUM", or "HIGH"
}`;

        const response = await this.makeRequest(prompt, { temperature: 0.4 });

        if (!response) return null;

        try {
            return JSON.parse(response);
        } catch (error) {
            return { recommendation: 'SKIP', reasoning: response };
        }
    }
}

export default GeminiClient;
